### ISSUE TITLE
Add a flag to strip potentially sensitive data from reports

### ISSUE THREAD

**Issue by ScottHelme** [w3c/webappsec-csp#111]
CSP reporting is incredibly useful and provides heaps of valuable data, but it can be difficult for smaller sites to implement. This was my main motivating factor in building https://report-uri.io, a free CSP reporting service. 

In the course of running this service I've come to terms with lots of odd browser behaviours and created a fairly robust set of filters to eliminate both noise and potentially sensitive information in reports. 

![filters](https://cloud.githubusercontent.com/assets/5352840/18003451/87d5e638-6b85-11e6-8d16-09075423d1a7.png)

I recently came across an issue where a report was being sent from a page that contained sensitive information in the URL as a query string. As the recipient of the report, report-uri.io had access to this sensitive information. With the above filters the query string is stripped at my edge and never hits storage, nor do I inspect or use the data for any other purposes, but it would be nice if the host had the ability to limit the information sent in the report. As more and more sites use CSP reporting, and CSP reporting as a service increases in popularity, this issue seems like it will only get bigger. 

Could we have some way of limiting the data sent in a report? The first thing that jumps to mind is the ability to strip the fragment and/or query string from locations like the referrer, document-uri and blocked-uri. We could be flexible and even strip the path or provide the option to not send the referrer at all. 

Given the nature of CSP reporting I don't think we can work on the assumption that reports will be sent to an endpoint under the control of the organisation.

**Comment by annevk** [comment]
If we do anything here we'd have to find a way to hide the URL entirely. Otherwise we're not really solving the capability URL problem (e.g., GitHub Gists, private GitHub repositories).

**Comment by nico3333fr** [comment]
It is a good idea, I've also found some bugs on some browsers that clearly transfer some informations that webmasters users should NOT be aware of: in some cases, I can know the visitor is using Ghostery (!) or some other extensions.
It could be also interesting to motivate people to choose secured URLs for report-uri. (it can be too bad that sensitive informations could be sent in clear by the browser)

**Comment by ScottHelme** [comment]
@annevk which URL would that be, the document-uri? If you didn't want a URL reporting in the document-uri would the answer not be to disable reporting on that particular page? The report could be close to useless without information on which page it happened on, no?

@nico3333fr I like the idea of using a secure URL for the report-uri directive, but I'm not sure how this could be enforced. It's down to the host to specify the endpoint so I guess the best we could do is pop an error in the dev console warning that the report-uri directive is not secure?

**Comment by annevk** [comment]
@ScottHelme maybe there's some way to encrypt it in a way that only the user agent and the origin in question can get back to it. Seems like you'd especially want to know about exploits on URLs protected through some capability.

**Comment by jonathanKingston** [comment]
@ScottHelme Is there much value in the reports still if the path is also obscured as @annevk is suggesting?

I agree with @nico3333fr that having only https reporting would have been a simpler solution for this. Site owners should really trust who they are sending their reports to really.

Could the URI or path be specified in the CSP? Some form of unique reference to replace the normal URI?

**Comment by ScottHelme** [comment]
@jonathanKingston I see your point, but this is already becoming a growing problem for report-uri.io and CSP deployment, let alone reporting, is minimal. 

Imagine if the query string parameters contain sensitive data like a name, address, maybe card data or a password reset token. Yes, the site in question trusts me enough to send me reports but I most definitely don't want that kind of data and the ability to neutralise it at source rather than rely on me to scrub it is much preferred. 

There's also many other ways that this could cause a problem like local laws, regulations or for compliance reasons. If the data isn't required to be sent, then I think we should have a way to not send it.

**Comment by jonathanKingston** [comment]
@ScottHelme agreed although I would suspect those sites are in violation of the laws anyway.

Is the unique token or encryption for the URI an option? As @annevk said the data can be embedded in the URL itself just the same. Perhaps that adds too much complexity anyway.

Giving the site owner the ability to filter down to path and/or hostname might solve most cases but then I don't see a way for a site owner to easily marry the report back to a specific problem without the token/encryption.

**Comment by ScottHelme** [comment]
@jonathanKingston I'm not sure I'm following what you mean so excuse me if my response is a little whacky!

As for encrypting, do you mean if the browser encrypted the string and I then simply provide it back to the host and they can decrypt? If so, that makes the job of the service incredibly hard as they can't do any filtering, searching, or querying based on that field. 

Take the following report:

    {csp-report:{
        "document-uri":"https://scotthelme.co.uk/passwordReset?token=abc123"
        ...
    }

If I were to send that in a CSP report then whoever owns the password reset token is at risk. If I can strip that out of the report then the report is still just as useful and we've protected the user that owns the token. Obviously there are scenarios when the query string may be needed but giving the host an option to remove it from reports would be nice!

**Comment by jonathanKingston** [comment]
@ScottHelme it was @annevk who was suggesting the encryption however it makes sense if we actually want to get the data back. Which for the likes of Single Page Apps and other modern sites the query string could really help cut down the problem area when debugging.

However I was additionally suggesting perhaps we can get reports like:
```
{csp-report:{
    "document-uri":"https://scotthelme.co.uk/passwordReset"
    ...
}
```

and
```
{csp-report:{
    "document-uri":"https://scotthelme.co.uk"
    ...
}
```

But also:
```
{csp-report:{
    "document-uri":"https://scotthelme.co.uk",
    "unique-token": "2422ad858e65cd7ad7d87a9c47a7517777ec6971"
    ...
}
```
Or:
```
{csp-report:{
    "document-uri":"https://scotthelme.co.uk",
    "encrypted-uri": "881504c9eef48e9d389ffd25c164e8e84ce157e3abc6f72ae3a5a69ee6e30317..."
    ...
}
```

(Sorry for the length of this)

**Comment by ScottHelme** [comment]
@jonathanKingston The only problem with the encryption aspect is the added complexity. You'd need to deliver a key in the policy and have the browser do the crypto before POSTing the payload. Perhaps there would be performance concerns there too. 

I like the look of the first report, couldn't we achieve that with a simple directive like strip-parameters? If the site needs the query string params then the directive could be included only on sensitive pages to cut down leakage.

**Comment by jonathanKingston** [comment]
I was hoping for both truncation solutions so you could strip what you needed. I suspect there would need to be a key so you could be aware that it's truncated too:

```
{csp-report:{
    "document-uri":"https://scotthelme.co.uk",
    "truncated-to": "hostname"
    ...
}
```

Thinking about the encryption aspect, sites that are of that size and experience would likely just encrypt on their own servers, so it's probably too complex for the usage it would get... perhaps report-uri enterprize solution for those companies😉

**Comment by ScottHelme** [comment]
You could do so with multiple directives I guess, strip-(fragment|query|path). 

I don't think an extra field in the report is necessary as the browser already returns original-policy so you could see the directives in that to determine if/how it was stripped.

**Comment by estark37** [comment]
People (I can't remember who, sorry) have floated the proposal that Referrer Policy should control URLs sent in CSP reports. I'm not a huge fan of that proposal, and it also wouldn't solve this particular problem (since RP doesn't give you a way to strip just query string), but it seems related so I thought I'd bring it up.

@ScottHelme, I wonder if you've thought about distributing a JS library that listens for SecurityPolicyViolationEvents and does the reporting manually, stripping out any information you don't want? Though that would be a lot more friction for your users.

**Comment by jacobbednarz** [comment]
❤️  the intent here.

I'm not sold on using encryption as I think that would be yet another hurdle for the majority of users to get a content security policy enabled. This would also be a pain to manage when you need to dig into the report and decrypt the values it sent through. 

> I agree with @nico3333fr that having only https reporting would have been a simpler solution for this. Site owners should really trust who they are sending their reports to. 

@jonathanKingston In this context, HTTPS is irrelevant as unless it's encrypted at rest, the transport mode isn't going to save it from being compromised or leaked once the provider has it. The motto of "Easiest way to not leak sensitive material is to not handle it" comes to mind here so I think it's great service providers like @ScottHelme are raising this instead of just trying to scrub it out.

> People (I can't remember who, sorry) have floated the proposal that Referrer Policy should control URLs sent in CSP reports. I'm not a huge fan of that proposal, and it also wouldn't solve this particular problem (since RP doesn't give you a way to strip just query string), but it seems related so I thought I'd bring it up.

This sounds like one solution however not everyone is using a referrer policy so I think this may leave us in the same position and like @estark37, I'd rather keep these two concerns separate.

The only way I could see this being able to be implemented is using something like @ScottHelme's suggestion of a `strip-(fragment|query|path)` option that would apply to the field in question and could be altered on a page-by-page basis where needed. Perhaps a default of `strip-query` would be suitable for most of the use cases and should you need further debugging on a page, you could conditionally remove it? This would solve the issue of people being unaware that they are sending this sensitive information to a reporting service.

**Comment by ScottHelme** [comment]
@estark37 I have thought about it yes but I want to make the barrier to entry as low as possible. This is one of the reasons I don't like not having a reporting mode flag in the CSP report itself so I can auto detect whether the policy was enforced or not. You'd be amazed at how many people switch from report only to enforced and then email me saying the service isn't showing their reports when they simply haven't updated the reporting address! I don't think we should push that on to the user ¯\_(ツ)_/¯

@jacobbednarz you couldn't be more accurate: "Easiest way to not leak sensitive material is to not handle it". I'm confident in my ability to strip this at my edge when the filter is set but the simple truth is I just don't want to. I don't want someone to have to think 'can I trust the service provider' when they don't have or need to. It should be down to the host what information they send me, not down to me to filter things they don't want me to have. 

Thinking about this a little more would these restrictions only need to be applied to the` 'document-uri'` and `'blocked-uri'` field? It's down to the referring site to worry about the content of the `'referrer'` field but I guess the `'blocked-uri'` could also be same origin and contain sensitive GET params so perhaps we should include both of those? 

    Content-Security-Policy: default-src 'self'; report-uri https://scotthelme.report-uri.io/r/default/csp/enforce; report-filter strip-fragment strip-query (strip-path)

**Comment by shekyan** [comment]
Users that put sensitive data in query string have all the possibilities to create their own [`SecurityPolicyViolationEvent`](https://w3c.github.io/webappsec-csp/#securitypolicyviolationevent), copy/filter whatever they are interested to do with it's properties and report on their own.

**Comment by ScottHelme** [comment]
@shekyan they do have that option but it greatly increases the barrier to entry and will result in a reduction of people using reporting. By the same logic, we could do away with CSP reporting by the browser completely and just ask that users implement their own library specific to their own needs. The addition of this simple directive is a backwards compatible measure that reduces friction to basically nothing. I'd rather we provide the functionality to users than have them write and maintain their own library and have the overhead of serving it along with every page load.

**Comment by ScottHelme** [comment]
Just a quick thought about the above example, you wouldn't need to specify multiple tokens in the theoretical report-filter directive. The strip-fragment would do as expected, strip-query would also imply strip-fragment and then strip-path would also imply both strip-query and strip-fragment.

**Comment by ScottHelme** [comment]
I'm now waiting for @mikewest to return and tell me why this is a terrible idea ;-)

**Comment by jacobbednarz** [comment]
Only drawback I see here is the extra values in an already long HTTP header but ¯\_(ツ)_/¯

**Comment by ScottHelme** [comment]
On the bright side they're only there for people who need them. If you need them then I guess you won't mind!

**Comment by mikewest** [comment]
It sounds like the threat that you care about here is leakage of the URL of the document in which a violation occurred to a third-party reporting endpoint. This sounds to me to be substantially similar to the kinds of things that we worry about with regard to referrers.

If you're worried about leaking query string data to a third-party, it seems like you should be worried about it for things other than CSP. That is, adding a toggle to the policy to govern its URL-stripping behavior solves only a (small) piece of the problem. Unless you change the referrer policy for the page, you're still at risk for leaking data to any third-party endpoint, period. Is it reasonable to want a distinct policy for CSP violation reports? Is there a case in which you'd care about locking one down but not the other?

Naively, it seems to me that we could simply apply the page's referrer policy to the URL with respect to the reporting endpoint. If that doesn't give enough granularity, perhaps we should add more toggles to referrer policy.

@estark37: You suggested above that you're not thrilled with that model; could you elaborate a bit? It seems like a pretty reasonable fit to me (though, admittedly I haven't thought about referrer policy as much as you have).

**Comment by ScottHelme** [comment]
You're right, given enough granularity in the referrer policy and having that applied to CSP reports would be sufficient for this scenario. I can't think of a case where you might want to lock one down and not the other but I will think it over some more. This seems like a good option.

**Comment by mikewest** [comment]
Ok. I think @estark37 is out of office for a bit; I'll wait for her to chat with @jeisinger about whether they see this as appropriate use of referrer policy. If they're on board, I'll update the CSP spec.

**Comment by estark37** [comment]
Hello! I'm computer-less for another week so I might be mildly incoherent
on this tiny keyboard, but, briefly, the gist of my uneasiness is that I
would think that developers would be confused to find that referrer policy
controls URLs in CSP reports but not other sources of leakage like Origin
headers and HPKP reports. I would hate for someone to get the impression
that, say, a referrer policy of 'no-referrer' strips Origin headers when it
doesn't.

But I can also see an argument that CSP reports are what bite people the
hardest and most often, and that devs would like RP to control CSP reports
and so it should just work that way. So I wouldn't be all that sad either
way.

**Comment by ScottHelme** [comment]
@estark37 for HPKP I think it's just the hostname that gets sent in the report, I don't think the path or query string make it in according to the spec? Perhaps @mikewest can confirm, I'm going from memory here. Of course, this applies to the Referer header of the POST request if that's what you meant though.

**Comment by estark37** [comment]
Yeah it's just the hostname in the hpkp report, but I'm thinking that if a
RP of 'no-referrer' strips all referrer info from csp reports, one might
expect it to strip the hostname from hpkp reports too.

**Comment by ScottHelme** [comment]
@estark37 hmm, that could be a problem then. A HPKP report or CSP report with no information in the hostname or document-URI field is almost useless isn't it? The host may not want to send referrers to 3rd party sites on outbound links but still needs the info in their reports .
